### DISCUSSION TITLE
How to return 404 or 500 or redirect in Next.js 13 custom hook for a page?

### DISCUSSION THREAD

**Discussion by fmotlagh03** [vercel/next.js#45466] [Help] [answered]
In Next.js 13 we can fetch data in each component, page or layout. This is great.

Now, we want to reuse our data fetching logic. Thus we have created a custom hook.

We want to reuse the logic of loading blog data across many projects. This is our code:

```
import { useBlogData } from 'blog'

const Blog = async () => {

    const { latestPosts, categories, tags, authors } = await useBlogData()

    return <div>Blog HTML</div>
}
```

As you can see, our `useBlogData` has a complex logic to support different queries for tags, categories, authors, pagination, SEO, etc.

However, we don't know how to return 404 or 500 or how to redirect **from inside** it.

How can we do that?

For example, if a post does not exist, how can we return 404?

**Comment by icyJoseph** [comment, marked as answer]
Hi,

You can call the `notFound` function for 404.

https://beta.nextjs.org/docs/api-reference/notfound#notfound

It'll cause the rendering to throw and it'll be caught by the SSR process to show a 404 page.

For redirect, there's a function called `redirect` as well:

https://beta.nextjs.org/docs/api-reference/redirect

Both of the above just throw an specific type of error, be it NEXT_NOT_FOUND or NEXT_REDIRECT, which the SSR process uses to decide what to do when SSR throws an exception.

Your `useBlogData` hook could return the response status, and you could easily check for 2xx, 3xx, or 4xx, otherwise, just check if your data is present.

For 5xx is a bit more tricky I think. You'll need an error page, https://beta.nextjs.org/docs/api-reference/file-conventions/error, and if 5xx has been hit, throw an error, but I am not sure if there's a custom error, or if a `throw new Error("boom")` is sufficient. I am not 100% sure on how to yield a 500 just yet.

**Reply by fmotlagh03** [reply]
Thank you so much
